### DISCUSSION TITLE
`impl Widget for &W` v.s. `impl WidgetRef for W`

### DISCUSSION THREAD

**Discussion by rhysd** [ratatui/ratatui#1274] [Q&A] [answered]
ratatui recently introduced [`WidgetRef`](https://docs.rs/ratatui/latest/ratatui/widgets/trait.WidgetRef.html) trait under the feature flag.

The document says

> Implementors should prefer to implement this over the Widget trait

However I'm not sure the advantage of the trait against implementing `Widget` for references. What is the advantage to introduce `WidgetRef` trait over just implementing `Widget` for references?

More specifically, let's say we have a widget type `W`, then

what is the advantage of

```rust
impl WidgetRef for W {
  fn render_ref(&self, area: Rect, buf: &mut Buffer) {
    // ...
  }
}
```

against

```rust
impl Widget for &W {
  fn render(self, area: Rect, buf: &mut Buffer) {
    // Here `self` is `&W` so it does not move out the widget instance.
  }
}
```

Adding a new trait increases the complexity of the crate. So I believe there is some advantage against a solution which introduces no new trait. I'd like to understand it.

**Comment by orhun** [comment]
Related discussion on Rust forums: <https://users.rust-lang.org/t/calling-a-method-with-a-self-parameter-on-box-dyn-trait/106220>

Also see #903 for the rationale. Of course, @joshka would provide a better explanation here :)

**Reply by rhysd** [reply]
Thanks. I missed the discussion on the forum.

**Reply by joshka** [reply]
I've added a tracking issue that pulls together links to the context of this more thoroughly at https://github.com/ratatui-org/ratatui/issues/1287 and linked to it from the docs.

**Comment by joshka** [comment, marked as answer]
TLDR:
- `Widget` probably should have been implemented as a method accepting `&self` from the outset, so `WidgetRef` is a more useful `Widget` trait that allows non consuming use.
- `WidgetRef` is still unstable as we're working out whether to keep the name `WidgetRef` or use `Render` instead. I didn't anticipate that we'd keep it unstable this long
- `WidgetRef` is required rather than just implementing `Widget for &T` when you want `T` to be any possible boxed widget (as the box cannot consume the contents.
- Implementing `Widget for &T` previously was pretty difficult, as your implementation had to create any child widgets and consume them on render. We changed all the built-in widgets to be non-mutating and implemented Widget for &T on them all which makes this easier now.
- If your needs are relatively simple then implementing `Widget for &T` (e.g. your text area widget) is probably the right thing to do.

(We should update the docs with this info)

**Reply by rhysd** [reply]
Thank you for the details.

My understanding is that the point is usefulness in terms of widget implementors and the ability between `impl Widget for &W` and `impl WidgetRef for W` is the same (at least for now), am I correct? Implementing `Widget` for references not to move out the instance is a bit trickier and harder to come up with. The `WidgetRef` trait is specific to the use case and is more explicit way.

> We changed all the built-in widgets to be non-mutating and implemented Widget for &T on them all which makes this easier now.

I can feel that this was a tough work. Thank you for the work and I think it is the right direction.

> WidgetRef is required rather than just implementing Widget for &T when you want T to be any possible boxed widget (as the box cannot consume the contents.

Would you elaborate on this? I think `Box<W>` can be handled by `impl Widget for &W` as follows thanks to [auto dereferncing](https://doc.rust-lang.org/std/ops/trait.Deref.html). So I could not understand why `WidgetRef` is required here.

```rust
let boxed_widget = Box::new(W { /* ... */ });

// ...

// `&box_widget` is the same as `boxed_widget.deref()` and returns &W reference
f.render_widget(&boxed_widget, f.size());
```

> If your needs are relatively simple then implementing Widget for &T (e.g. your text area widget) is probably the right thing to do.

Yes, currently tui-textarea's `TextArea` widget implements `Widget` for `&TextArea<'_>`. Do you think I should migrate it to implement `WidgetRef` for `TextArea<'_>` considering the future of ratatui? If so, I will do it once the `WidgetRef` trait is stabilized.

**Reply by joshka** [reply]
> Would you elaborate on this? I think `Box<W>` can be handled by `impl Widget for &W` as follows thanks to [auto dereferncing](https://doc.rust-lang.org/std/ops/trait.Deref.html). So I could not understand why `WidgetRef` is required here.
> 
> ```rust
> let boxed_widget = Box::new(W { /* ... */ });
> 
> // ...
> 
> // `&box_widget` is the same as `boxed_widget.deref()` and returns &W reference
> f.render_widget(&boxed_widget, f.size());
> ```

`boxed_widget` there is `Box<W>`, not `Box<dyn Widget>`. It can only hold a single type of widget, not all types. The heterogeneous part of this is where this becomes useful, not the boxed part 

Consider:
```rust
    let label = FormLabel::new("label");
    let text = FormText::new("text");
    let widgets: Vec<Box<dyn Widget>> = vec![Box::new(&label), Box::new(&text)];
    terminal.draw(|frame| {
        for (widget, row) in zip(widgets, frame.size().rows()) {
            frame.render_widget(widget, row);
        }
    })?;
```
<details><summary>With supporting code:</summary>

```rust
struct FormLabel {
    text: String,
}

impl FormLabel {
    fn new(text: &str) -> Self {
        Self {
            text: text.to_string(),
        }
    }
}

impl Widget for &FormLabel {
    fn render(self, area: Rect, buf: &mut Buffer) {
        Line::raw(self.text.clone()).render(area, buf);
    }
}

struct FormText {
    text: String,
}

impl FormText {
    fn new(text: &str) -> Self {
        Self {
            text: text.to_string(),
        }
    }
}

impl Widget for &FormText {
    fn render(self, area: Rect, buf: &mut Buffer) {
        Line::raw(self.text.clone()).render(area, buf);
    }
}
```

</details> 

This requires Widget to be implemented for `Box<dyn Widget>` which isn't really possible easily afaik. The closest I can get is:

```rust
impl<T> Widget for Box<T>
where
    T: Widget + Copy + Sized,
{
    fn render(self, area: Rect, buf: &mut Buffer) {
        self.deref().render(area, buf);
    }
}
```

But this doesn't quite solve the problem. There's probably a way to do this (perhaps with higher order traits, or something similar), but WidgetRef was the one thing I found that actually worked for this use case. Perhaps you have some more insight here.

> Do you think I should migrate it to implement `WidgetRef` for `TextArea<'_>` considering the future of ratatui?

Probably not for now - keep it simple until you see a use case that makes this a problem. We're still experimenting on this and understanding it a bit better. I really should change the docs on this.

**Reply by joshka** [reply]
Updated the docs guidance around WidgetRef in https://github.com/ratatui-org/ratatui/pull/1288

**Reply by rhysd** [reply]
Ah, I forgot the trait objects. Thank you for the detailed explanation. That's much helpful for me.

> Probably not for now - keep it simple until you see a use case that makes this a problem.

Sure. Then I'll stick with the current implementation until some issue is reported.

> Updated the docs guidance around WidgetRef in https://github.com/ratatui-org/ratatui/pull/1288

Thank you for the clarification.

**Comment by joshka** [comment]
As a followup to this, in https://github.com/ratatui/ratatui/pull/1491 we're changing widgets to be a blanket implementation of WidgetRef for any Widget implementation on &W, rather than the other way around (Widget for WidgetRef). This makes more sense and solidifies the idea that WidgetRef is more of a supporting type than the main type.

**Comment by ArtyomSinyugin** [comment]
Am I correct in understanding that the Widget trait in its current implementation exists primarily because changing it would break other projects? Otherwise, it would be more convenient to make WidgetRef the main implementation? That is, rename the current WidgetRef to Widget and remove the old implementation?

**Reply by joshka** [reply]
Widget is generally usable. It likely should have accepted `&self` originally and perhaps had a WidgetMut version, but that's not the end of the world. Implement it on `&YourType` or `&mut YourType` and forget that it takes ownership. WidgetRef then only really exists for places where this ownership is difficult to use (boxed dynamic types).

See https://github.com/ratatui/ratatui/tree/main/examples/concepts/state and https://docs.rs/ratatui/0.30.0-beta.0/ratatui/widgets/ for more info (the 0.30-beta docs have a significant amount of detail that's not in the current docs).
